### PR TITLE
[impeller] OpenGL: Add support for threads and contexts.

### PR DESCRIPTION
**Pull request by chinmaygarde**
I have temporarily made GL errors log instead of assert. But the only issues we
run into are with the "create on bind" resources and setting debug object labels
on them. I am going to rework setting debug labels right now. But, in the
meantime, the OpenGL ES backend can show cross-context textures.

There is also still an outstanding issue where cross-context textures will be
uploaded on the main context (in the same sharegroup) if the thread on which the
operation is dispatched doesn't have a context current. This will also be fixed
to pin those tasks to the worker context in an upcoming patch.
screen

### CODE REVIEW

**Comment by chinmaygarde** [comment]
The screenshot reminds me, there is no mip-mapping just yet either.

File: impeller/renderer/backend/gles/reactor_gles.cc
Comment (dnfield) [inline impeller/renderer/backend/gles/reactor_gles.cc:197]: Is this runtime asserting the behavior of swap?

**Review by dnfield — APPROVED** [review]
Change seems good, it'd be good to make sure this builds in google3 before landing

File: impeller/renderer/backend/gles/reactor_gles.cc
Comment (chinmaygarde) [reply impeller/renderer/backend/gles/reactor_gles.cc:197]: Umm, yeah. Thats possibly a bit much. I'll get rid of it in a subsequent patch.

**Comment by chinmaygarde** [comment]
> it'd be good to make sure this builds in google3 before landing

Is there anything in particular that seems unique about this patch vis-a-vis building in g3? I thought this was much like the rest of the work on ES2.

**Comment by dnfield** [comment]
impeller isn't getting rolled into google3, so I added some compiler guards around other code that references impeller classes/headers. You seem to have added some more in this patch, specifically to files that don't match `*_impeller.*` in the Android folder - that may cause compilation to fail internally.

**Comment by dnfield** [comment]
(We're in a world right now where the android engine needs to be able to build as if the `impeller` directory does not exist and no build files refer to anything in it or anything with `_impeller` in the file name)

**Comment by chinmaygarde** [comment]
The two files that are not _impeller or in the impeller/ dep are using forward declarations already. So I think we are good.

**Comment by zanderso** [comment]
Worst frame time started to vary wildly after this PR:

https://flutter-flutter-perf.skia.org/e/?begin=1652929632&end=1653452589&queries=device_type%3DMoto_G__4_%26sub_result%3D90th_percentile_frame_rasterizer_time_millis%26sub_result%3D99th_percentile_frame_rasterizer_time_millis%26sub_result%3Daverage_frame_rasterizer_time_millis%26sub_result%3Dworst_frame_rasterizer_time_millis%26test%3Dnew_gallery__transition_perf%26test%3Dnew_gallery_impeller__transition_perf&requestType=0

It sounds like that might be expected. Is this addressed by https://github.com/flutter/engine/pull/33598 ?

**Comment by chinmaygarde** [comment]
The issue with the varying worst frame times is because the texture might be uploaded using the main context instead of the resource context. See the caveat in the original commit message:
```
There is also still an outstanding issue where cross-context textures will be
uploaded on the main context (in the same sharegroup) if the thread on which the
operation is dispatched doesn't have a context current. This will also be fixed
to pin those tasks to the worker context in an upcoming patch.
```
https://github.com/flutter/engine/pull/33598 doesn't address this but the commit right after should as that patch sets up the necessary refactors for addressing pinned tasks. Rest assured, the regressions are due to images actually being uploaded and displayed, just using the wrong context. If we didn't have a resource context in Flutter, thats what the benchmarks would look like.

I'll keep a lookout on the benchmarks once we call the OpenGL ES backend work done for this milestone. Thanks.

**Pull request merged**
